Sentry blog article (#404)
quarkusio/quarkusio.github.io: _posts/2019-12-18-sentry-the-guardian-of-your-quarkus-app.adoc

---
layout: post

> title: 'Sentry Error Tracker - The new guardian of your Quarkus application'

Reviewer: minor nit but due to sentry's recent license changes it rings wrong to me that we call this "*the* new guardian of your quarkus application" ....could we tone it down to for example "A guardian of your Quarkus application" . ?

*This is awesome!!*

Mmmeahh *you may have forgotten a detail...* have you thought about *the fate of your first production exception* (please be honest with me :-)?

["upperalpha"]
. As a backend god, master of the world, first of his name, you implemented some logstash process to extract your logs files, send them to some elastic search server somewhere and... you just spend more time configuring it than actually coding your app 😂.
. You thought you could manually check the error logs a few times a week and ... oups, so unlucky, your application broke exactly the day you forgot to check, let's hope no one else noticed 😅 ?
. *You were waiting for this article 🤓*

Now that you are teased, aren't you? Let's dive in. 

== Sentry, the guardian has arrived on Quarkus

Once your code is reviewed, tested in many ways, containerized, ..., and finally *promoted to production*, [line-through]#I guess you can expect it to run smoothly ever after..# *an exception is going to pop*, it's just a question of time. The questions are:

> 

Reviewer: will have to mention their recent license changes so we don't end up recommending this as something to be usable in opensource blindly.

Author: Yes, thanks for the info @maxandersen, Good catch !!

. *How will you get notified about this error?*
. *What are you going to do about it?*

Well, this looks like a complex problem right? This is what I really like about using *Quarkus and Sentry together*, you will be all set in a few minutes:

> * Create a https://sentry.io/signup/[Sentry account], you can choose a free plan or another depending on your project needs (or with a bit more time, even a https://docs.sentry.io/server/[self hosted sentry])

Reviewer: could you add something like "Be aware sentry.io recently changed to a non-opensource license" or similar as if you just read this you could read it as you can just use it like any other opensource project we link/refer to.

* Configure Sentry in your Quarkus application
. Add the extension:
+
```
$ ./mvnw quarkus:add-extension -Dextensions=sentry
```
. Set it up: 
+
```
quarkus.log.sentry=true
quarkus.log.sentry.dsn=[your Sentry DSN]
quarkus.log.sentry.in-app-packages=org.acme
```
+ 

> NOTE: More details about the configuration can be found in the https://quarkus.io/guides/logging-sentry[Logging with Sentry guide]

Reviewer: we shouldn't post this before that link will work.

Reviewer: also would be good if https://github.com/quarkusio/quarkus/pull/6302 could be included as the doc in master states sentry.io is opensource.

---
layout: post
title: 'Sentry Error Tracker - The new guardian of your Quarkus application'

> date: 2019-12-18

Reviewer 2: You should probably adjust the date once you are done with the adjustments (header + file name).

tags: quarkus extension logging cloud error
author: adamevin
---

:imagesdir: /assets/images/posts/quarkus-sentry

> *If like me you have met Quarkus and fell in love*, you started implementing your backend and you started to feel so light and quick, click clack, whoot whoot... *in a woosh, your lightning fast native container is ready to be sent on your brand new kubernetes cluster 🤙*

Reviewer 2: ```suggestion
*If like me you have met Quarkus and fell in love*, you started implementing your backend and you started to feel so light and quick, click clack, whoot whoot... *in a woosh, your lightning fast native container is ready to be sent on your brand new Kubernetes cluster 🤙*
```

*This is awesome!!*

Mmmeahh *you may have forgotten a detail...* have you thought about *the fate of your first production exception* (please be honest with me :-)?

["upperalpha"]

> . As a backend god, master of the world, first of his name, you implemented some logstash process to extract your logs files, send them to some elastic search server somewhere and... sorry to say it, but you just spend more time configuring it than actually coding your app 😅.

Reviewer 2: ```suggestion
. As a backend god, master of the world, first of his name, you implemented some Logstash process to extract your log files, send them to some Elasticsearch server somewhere and... sorry to say it, but you just spend more time configuring it than actually coding your app 😅.
```

> . You thought you could manually check the error logs a few times a week and ... oups, so unlucky, your application broke exactly the day you forgot to check, let's hope no one else noticed 🤭?

Reviewer 2: ```suggestion
. You thought you could manually check the error logs a few times a week and ... oops, unlucky you, your application broke exactly the day you forgot to check, let's hope no one else noticed 🤭?
```

. *You were waiting for this article 🤓*

Now that you are teased, aren't you? Let's dive in. 

== Sentry, the guardian has arrived on Quarkus

Once your code is reviewed, tested in many ways, containerized, ..., and finally *promoted to production*, [line-through]#I guess you can expect it to run smoothly ever after..# *an exception is going to pop*, it's just a question of time. The questions are:

. *How will you get notified about this error?*

> . *What are you going to do about it?*

Reviewer 2: I'm not sure I would use italic everywhere in this bullet list? Looks like a bit too much italic for our sake :).

Author: It's supposed to be bold, no?

Reviewer: agree with @gsmet here - all the boldness distracts from the actual content of the message imo.

![2019-12-21_08-54-57](https://user-images.githubusercontent.com/54129/71305137-926c6080-23cf-11ea-8808-a6159b3ad6be.png)

I would say you don't need the visual gimmick as the content is good on its own.

Well, this looks like a complex problem right? This is what I really like about using *Quarkus and Sentry together*, you will be all set in a few minutes:

* Create a https://sentry.io/signup/[Sentry account], you can choose a free plan or another depending on your project needs (or with a bit more time, even a https://docs.sentry.io/server/[self hosted sentry])

> * Configure Sentry on your Quarkus application

Reviewer 2: ```suggestion
* Configure Sentry in your Quarkus application
```

. Add the extension:
+

> ```

Reviewer 2: You should use proper `[source]` for Asciidoc (even if it somehow works with the Markdown one).

Same remark for all the source extracts.

$ ./mvnw quarkus:add-extension -Dextensions=sentry
```
. Set it up: 
+
```
quarkus.log.sentry=true
quarkus.log.sentry.dsn=[your Sentry DSN]
quarkus.log.sentry.in-app-packages=org.acme
```
+ 

> NOTE: More details about configuration can be found in the https://quarkus.io/guides/logging-sentry[guide]

Reviewer 2: ```suggestion
NOTE: More details about the configuration can be found in the https://quarkus.io/guides/logging-sentry[Logging with Sentry guide]
```
(the latter change is just to improve SEO)

You now enter the really pretty Sentry application. It is also smart enough to *group similar events into issues* (or you can also decide to group them yourself). 

image::sentry-issues.png[Sentry Issue List, 900]

It's now time to *generate and link an issue on your favorite tracker* (GitHub, Bitbucket, GitLab, ...). It will contain a lot of *contextual information about the errors* (endpoint, params, stacktrace, environment, ...) to help your team fix that nasty bug. 

image::sentry-issue.png[Sentry issue details, 900]

If you want to go a bit further you can start cooking some *custom rules to dispatch alerts* with your own taste.

image::sentry-alerts-config.png[Sentry Alerts Configuration, 900]

Bonus point, *Sentry is also available for your https://sentry.io/for/javascript/[frontend]*.

Good luck 🤙

NOTE: Once an extension is integrated in the Quarkus eco-system, the life of thousand of developers is made easier. If you know of a similar (or any) tool you think could help the Quarkus community, https://github.com/quarkusio/quarkus/issues/new?assignees=&labels=extension-proposal&template=extension_proposal.md&title=[propose it]!

> NOTE: Sentry has recently https://blog.sentry.io/2019/11/06/relicensing-sentry[re-licensed most of its code to BSL] - which is not an OSI-approved license. The Java SDK is still under the 3-Clause BSD license.

Reviewer 3: What's the impsct this has in practical terms?

Reviewer 2: It's the new "we don't like Amazon using our stuff without paying" type of license.

Basically:
- you have access to the source code and can modify it but there are restrictions about the usage - typically you may have to pay to use it in production under certain conditions
- the software will eventually become Open Source after a period of time (counted in years)

So that's definitely not an Open Source software anymore. And this is going to be a major pain.

Reviewer: IANAL, but yes, and it means a company like red hat which easily can be categorised as a company doing services that help trace errors upgrades to use latest sentry code on their own-hosted services they are in breach of license.  

in other words - to continue use sentry from our opensource projects using own hosted software we'll most likely have to use pre-november sourcecode hosted sentry.

Reviewer: Also, IMO that NOTE should be a WARNING and I would prefer it to be higher up towards the section on how to setup sentry account.
